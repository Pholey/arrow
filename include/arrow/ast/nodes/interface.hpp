// Copyright Â© 2014-2016 Ryan Leckey, All Rights Reserved.

// Distributed under the MIT License
// See accompanying file LICENSE

#ifndef ARROW_AST_INTERFACE_H
#define ARROW_AST_INTERFACE_H

#include <string>

#include "arrow/ptr.hpp"
#include "arrow/ast/nodes/name.hpp"
#include "arrow/ast/nodes/function.hpp"
#include "arrow/ast/nodes/statement.hpp"

namespace arrow {
namespace ast {

struct Interface : Statement {
  Interface(Span span, std::string name, std::vector<ptr<ast::TypeParameter>> type_parameters, std::vector<ptr<Function>> functions)
    : Statement(span), name(name), type_parameters(type_parameters), functions(functions) {
  }

  virtual ~Interface() noexcept;

  /// Name
  std::string name;

  /// Type Parameters
  std::vector<ptr<TypeParameter>> type_parameters;

  /// Functions
  std::vector<ptr<Function>> functions;
};

}  // namespace ast
}  // namespace arrow

#endif  // ARROW_AST_INTERFACE_H
