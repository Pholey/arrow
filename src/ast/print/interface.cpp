// Copyright Â© 2014-2016 Ryan Leckey, All Rights Reserved.

// Distributed under the MIT License
// See accompanying file LICENSE

#include "mach7.hpp"
#include "arrow/ast/print.hpp"

using arrow::ast::Print;

void Print::print_interface(ptr<Interface> n) {
  handle("Interface", n, [&, this] {
    _w.Key("type_parameters");
    _w.StartArray();

    for (auto& param : n->type_parameters) {
      print(param);
    }

    _w.EndArray();

    _w.Key("name");
    _w.String(n->name.c_str());

    _w.Key("functions");
    _w.StartArray();

    for (auto& fn : n->functions) {
      print(fn);
    }

    _w.EndArray();
  });
}
