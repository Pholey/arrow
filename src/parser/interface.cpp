// Copyright Â© 2014-2016 Ryan Leckey, All Rights Reserved.

// Distributed under the MIT License
// See accompanying file LICENSE

#include "arrow/parser.hpp"
#include "arrow/log.hpp"

using arrow::Parser;

auto Parser::parse_interface() -> ptr<ast::Interface> {
  // Expect: `interface`
  auto tok_begin = expect(token::Type::Interface);
  if (!tok_begin) return nullptr;

  // Expect: identifier
  auto id = parse_id();
  if (!id) return nullptr;

  // Check for `<` .. `>` to indicate a generic interface
  std::vector<ptr<ast::TypeParameter>> type_parameters;
  if (_t.peek()->type == token::Type::LessThan) {
    _t.pop();

    if (!handle_sequence<ast::TypeParameter>(
      &(type_parameters),
      std::bind(&Parser::parse_type_parameter, this),
      token::Type::GreaterThan
    )) return nullptr;

    // Expect: `>`
    if (!expect(token::Type::GreaterThan)) return nullptr;
  }

  // Expect: `{`
  if (!expect(token::Type::LeftBrace)) return nullptr;

  // Iterate and attempt to parse member functions ..
  std::vector<ptr<ast::Function>> functions;
  while (
    (_t.peek()->type != token::Type::RightBrace) &&
    (_t.peek()->type != token::Type::End)
  ) {
    // Parse: Function
    auto fn = parse_function(false);
    if (!fn) return nullptr;
    functions.push_back(fn);
  }

  // Expect: `}`
  auto tok_end = expect(token::Type::RightBrace);
  if (!tok_end) return nullptr;

  return std::make_shared<ast::Interface>(
    tok_begin->span.extend(tok_end->span), id->text, type_parameters, functions);
}
