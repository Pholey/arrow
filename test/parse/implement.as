import "./interface";

struct Cell {
  value: str;
}

struct Box<T> {
  value: T;
}

struct Pair<T, U> {
  first: T;
  second: U;
}

implement Cell {
  def row(): Row {
    // [...]
  }
}

implement<T> Box<T> {
  def unwrap(self): T {
    return self.value;
  }
}

implement Pair<bool, Box<bool>> {
  // [...]
}

implement<T, U> Sequence<T> for Box<U> {
  // [...]
}

implement<T> Resource for Box<T> {
  def release(self) {
    // [...]
  }
}
