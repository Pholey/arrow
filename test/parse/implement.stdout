{
  "tag": "Module",
  "span": "test/parse/implement.as:1-40",
  "statements": [
    {
      "tag": "Import",
      "span": "test/parse/implement.as:1:1-21",
      "source": "./interface",
      "alias": ""
    },
    {
      "tag": "TypeRecord",
      "span": "test/parse/implement.as:3-5",
      "name": "Cell",
      "type_parameters": [],
      "members": [
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/implement.as:4:3-13",
          "name": "value",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/implement.as:4:10-13",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/implement.as:4:10-13",
              "text": "str"
            }
          }
        }
      ]
    },
    {
      "tag": "TypeRecord",
      "span": "test/parse/implement.as:7-9",
      "name": "Box",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:7:12-13",
          "name": "T"
        }
      ],
      "members": [
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/implement.as:8:3-11",
          "name": "value",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/implement.as:8:10-11",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/implement.as:8:10-11",
              "text": "T"
            }
          }
        }
      ]
    },
    {
      "tag": "TypeRecord",
      "span": "test/parse/implement.as:11-14",
      "name": "Pair",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:11:13-14",
          "name": "T"
        },
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:11:16-17",
          "name": "U"
        }
      ],
      "members": [
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/implement.as:12:3-11",
          "name": "first",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/implement.as:12:10-11",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/implement.as:12:10-11",
              "text": "T"
            }
          }
        },
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/implement.as:13:3-12",
          "name": "second",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/implement.as:13:11-12",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/implement.as:13:11-12",
              "text": "U"
            }
          }
        }
      ]
    },
    {
      "tag": "Implement",
      "span": "test/parse/implement.as:16-20",
      "type_parameters": [],
      "target": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:16:11-15",
        "text": "Cell"
      },
      "interface": null,
      "functions": [
        {
          "tag": "Function",
          "span": "test/parse/implement.as:17-19",
          "name": "row",
          "result_type": {
            "tag": "TypeName",
            "span": "test/parse/implement.as:17:14-17",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/implement.as:17:14-17",
              "text": "Row"
            }
          },
          "parameters": [],
          "type_parameters": [],
          "block": {
            "tag": "Block",
            "span": "test/parse/implement.as:17-19",
            "statements": []
          }
        }
      ]
    },
    {
      "tag": "Implement",
      "span": "test/parse/implement.as:22-26",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:22:11-12",
          "name": "T"
        }
      ],
      "target": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:22:14-20",
        "text": "Box"
      },
      "interface": null,
      "functions": [
        {
          "tag": "Function",
          "span": "test/parse/implement.as:23-25",
          "name": "unwrap",
          "result_type": {
            "tag": "TypeName",
            "span": "test/parse/implement.as:23:21-22",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/implement.as:23:21-22",
              "text": "T"
            }
          },
          "parameters": [
            {
              "tag": "Parameter",
              "span": "test/parse/implement.as:23:14-18",
              "name": "self",
              "type": {
                "tag": "TypeName",
                "span": "test/parse/implement.as:23:14-18",
                "name": {
                  "tag": "Identifier",
                  "span": "test/parse/implement.as:23:14-18",
                  "text": "Self"
                }
              }
            }
          ],
          "type_parameters": [],
          "block": {
            "tag": "Block",
            "span": "test/parse/implement.as:23-25",
            "statements": [
              {
                "tag": "Return",
                "span": "test/parse/implement.as:24:5-22",
                "operand": {
                  "tag": "Path",
                  "span": "test/parse/implement.as:24:12-22",
                  "operand": {
                    "tag": "Identifier",
                    "span": "test/parse/implement.as:24:12-16",
                    "text": "self"
                  },
                  "member": {
                    "tag": "Identifier",
                    "span": "test/parse/implement.as:24:17-22",
                    "text": "value"
                  }
                }
              }
            ]
          }
        }
      ]
    },
    {
      "tag": "Implement",
      "span": "test/parse/implement.as:28-30",
      "type_parameters": [],
      "target": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:28:11-32",
        "text": "Pair"
      },
      "interface": null,
      "functions": []
    },
    {
      "tag": "Implement",
      "span": "test/parse/implement.as:32-34",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:32:11-12",
          "name": "T"
        },
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:32:14-15",
          "name": "U"
        }
      ],
      "target": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:32:33-39",
        "text": "Box"
      },
      "interface": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:32:17-28",
        "text": "Sequence"
      },
      "functions": []
    },
    {
      "tag": "Implement",
      "span": "test/parse/implement.as:36-40",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/implement.as:36:11-12",
          "name": "T"
        }
      ],
      "target": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:36:27-33",
        "text": "Box"
      },
      "interface": {
        "tag": "Identifier",
        "span": "test/parse/implement.as:36:14-22",
        "text": "Resource"
      },
      "functions": [
        {
          "tag": "Function",
          "span": "test/parse/implement.as:37-39",
          "name": "release",
          "result_type": null,
          "parameters": [
            {
              "tag": "Parameter",
              "span": "test/parse/implement.as:37:15-19",
              "name": "self",
              "type": {
                "tag": "TypeName",
                "span": "test/parse/implement.as:37:15-19",
                "name": {
                  "tag": "Identifier",
                  "span": "test/parse/implement.as:37:15-19",
                  "text": "Self"
                }
              }
            }
          ],
          "type_parameters": [],
          "block": {
            "tag": "Block",
            "span": "test/parse/implement.as:37-39",
            "statements": []
          }
        }
      ]
    }
  ]
}
